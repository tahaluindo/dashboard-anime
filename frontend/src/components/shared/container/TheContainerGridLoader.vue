<template>
  <the-container-grid ref="container" v-if="items.length">
    <the-container-loader
      v-for="(item, index) in items"
      :key="index"
      class="opacity-0 flex flex-col"
      :data-delay="calcDelay(Number(index))"
    >
      <slot :item="item" />
    </the-container-loader>
    <the-button
      v-if="hasMore"
      class="
        btn-lg-border-md
        text-primary
        hover:text-white hover:bg-primary
        col-span-full
        font-opensans
        clamptext-08-12
        py-1
        justify-self-center
      "
      @click.prevent="loadMore"
    >
      Carregar mais items
    </the-button>
  </the-container-grid>
  <the-search-not-found v-else />
</template>

<script lang="ts">
import { defineComponent, PropType, ref } from "vue";
import { ElementObserver } from "./controller/elementObserver";
import TheContainerGrid from "./TheContainerGrid.vue";
import TheContainerLoader from "./TheContainerLoader.vue";
import TheButton from "../../shared/button/TheButton.vue";
import { getGridColumnsQty } from "../../../helpers/utils";
import TheSearchNotFound from "../search/TheSearchNotFound.vue";

export default defineComponent({
  components: {
    TheContainerGrid,
    TheContainerLoader,
    TheButton,
    TheSearchNotFound,
  },
  emits: ["loadMore"],
  props: {
    items: {
      type: Array,
      default: [],
    },
    hasMore: {
      type: Boolean,
      default: true,
    },
  },
  setup(_, { emit }) {
    const elementObserver = new ElementObserver();
    elementObserver.provide();

    const container = ref<HTMLElement>();

    const calcDelay = (index: number) => {
      const gridQantity = getGridColumnsQty();
      const pos = index % gridQantity;

      return pos * 100;
    };

    const loadMore = () => emit("loadMore");

    return {
      calcDelay,
      loadMore,
      container,
    };
  },
});
</script>

<style lang="postcss"></style>
