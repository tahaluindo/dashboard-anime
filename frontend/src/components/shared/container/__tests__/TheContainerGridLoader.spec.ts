import { mount } from "@vue/test-utils";
import {
  mockWindowIntersectionObserver,
  mockWindowVisualViewport,
} from "../../../../__mocks__/window.mock";
import TheContainerGridLoader from "../TheContainerGridLoader.vue";

describe("TheContainerGridLoader component test", () => {
  beforeEach(() => {
    mockWindowIntersectionObserver();
    mockWindowVisualViewport({ width: 1500 });
  });

  it("should render the default component", async () => {
    const wrapper = mount(TheContainerGridLoader, {
      shallow: true,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should call CalcDelay and this returns 0 to 0, 100 to 1 and 200 to 2", () => {
    const wrapper = mount(TheContainerGridLoader);

    const returnedValueZero = wrapper.vm.calcDelay(0);
    const returnedValueOne = wrapper.vm.calcDelay(1);
    const returnedValueTwo = wrapper.vm.calcDelay(2);

    expect(returnedValueZero).toBe(0);
    expect(returnedValueOne).toBe(100);
    expect(returnedValueTwo).toBe(200);
  });

  it("should emit 'loadMore' if loadMore function was called", () => {
    const wrapper = mount(TheContainerGridLoader);

    wrapper.vm.loadMore();

    expect(wrapper.emitted()).toHaveProperty("loadMore");
  });
});
