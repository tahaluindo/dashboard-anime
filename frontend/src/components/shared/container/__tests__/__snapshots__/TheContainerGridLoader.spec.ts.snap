// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheContainerGridLoader component test should render the default component 1`] = `"<the-search-not-found-stub></the-search-not-found-stub>"`;
